src/basic_solver.cpp

```diff
@@ -0,0 +1,166 @@
+    
+        // Calculate ATb contribution for this list entry
+        for (ssize_t o = 0; o < A_shape_one; o ++) {
+            ssize_t base_pos = o*N;
+            for (ssize_t l = 0; l < N; l++) {
+                ATb_data[base_pos+l] += K_data[l] * ATb_small(o);
+            }
+        }
+
+        // Calculate ATA contribution for this list entry
+        for (ssize_t j = 0; j < A_shape_one; j++) {
+            ssize_t base_pos_j = j*N;
+            for (ssize_t i = 0; i < j+1; i++) {
+                ssize_t base_pos = i*N;
+                ATA_eig.template block<N, N>(base_pos, base_pos_j) += ATA_small(i,j)*K_cross;
+            }
+        }
+    }
+
+    // solve Ax=b. If the matrix can be decomposed using llt do that (faster) if it cant be
+    // then use the slower ldlt method
+    Eigen::Map<const Eigen::Vector<T, Eigen::Dynamic>> ATb_map(ATb_data, A_shape[1]*N);
+    auto view = ATA_eig. template selfadjointView<Eigen::Upper>();
+    auto lltDecomp = view.llt();
+    Eigen::Matrix<T, Eigen::Dynamic, 1> result;
+    if (lltDecomp.info() == Eigen::ComputationInfo::Success) {
+        result = lltDecomp.solve(ATb_map);
+    } else {
+        auto decomp = view.ldlt();
+        result = decomp.solve(ATb_map);
+    }
+    return result;
+}
+
+template<class T>
+auto _solve_direct_cpp(
+        std::vector<py::array_t<T, py::array::f_style>> bList,
+        std::vector<py::array_t<T, py::array::f_style>> AList,
+        std::vector<py::array_t<T, py::array::f_style>> KList) {
+
+    // check some properties of the inputs, more will be done later in the
+    // implementation function
+    if (bList.size() < 1) {
+        throw std::runtime_error("Input data must have at least 1 element");
+    }
+    if (bList.size() != AList.size() || KList.size() != bList.size()) {
+        throw std::runtime_error("All inputs must be the same length");
+    }
+
+    py::buffer_info K_buffer_info = KList[0].request();
+
+    // Switch on the length of the K parameter, this templated value
+    // allows the compiler to make various optimizations
+    switch (K_buffer_info.shape[0]) {
+        case 1:
+            return _solve_direct_cpp_impl<T, 1>(bList, AList, KList);
+        case 2:
+            return _solve_direct_cpp_impl<T, 2>(bList, AList, KList);
+        case 3:
+            return _solve_direct_cpp_impl<T, 3>(bList, AList, KList);
+        case 4:
+            return _solve_direct_cpp_impl<T, 4>(bList, AList, KList);
+        case 5:
+            return _solve_direct_cpp_impl<T, 5>(bList, AList, KList);
+        case 6:
+            return _solve_direct_cpp_impl<T, 6>(bList, AList, KList);
+        case 7:
+            return _solve_direct_cpp_impl<T, 7>(bList, AList, KList);
+        case 8:
+            return _solve_direct_cpp_impl<T, 8>(bList, AList, KList);
+        case 9:
+            return _solve_direct_cpp_impl<T, 9>(bList, AList, KList);
+        case 10:
+            return _solve_direct_cpp_impl<T, 10>(bList, AList, KList);
```

Comment:
Ok, so I did a parallel branch of this branch where there is an additional cpp solver without template (`add_cpp_in_BasisInterp_check_template`)

I did some performance test on a single CCD from HSC with ~100 training PSF stars. The default Piff config: 


```
{'type': 'Simple',
 'model': {'type': 'PixelGrid',
  'scale': 0.16839226517851116,
  'size': 25,
  'interp': 'Lanczos(11)'},
 'interp': {'type': 'BasisPolynomial', 'order': 2, 'solver': 'scipy'},
 'outliers': {'type': 'Chisq', 'nsigma': 4.0, 'max_remove': 0.05}}
```

So here this is `case 6:`  and here are the difference in time to compute the PSF solutions for the different solver:

```
Solver = Scipy —> 20.770 seconds 
Solver = C++ (with template, case:6) —> 14.162 seconds
Solver = C++ (no template) —> 16.942 seconds 
```

I switched the order of the interpolation to 3 (`case: 10`, then) and here are the difference in time to compute the PSF solutions. 

```
Solver = Scipy —> 46.795 seconds
Solver = C++ (with template, case:10) —> 29.346 seconds
Solver = C++ (no template) —> 35.622 seconds 
```

So in both case, C++ is faster but this is even faster by using template. My understanding of why this is faster from discussion with Nate, this is coming from this block of lines:  

```
        // Calculate ATA contribution for this list entry
        for (ssize_t j = 0; j < A_shape_one; j++) {
            ssize_t base_pos_j = j*N;
            for (ssize_t i = 0; i < j+1; i++) {
                ssize_t base_pos = i*N;
                ATA_eig.template block<N, N>(base_pos, base_pos_j) += ATA_small(i,j)*K_cross;
                // this is where it was supposed to make the difference because of block assignement.
```

It allows a better implementation in each case at compilation time if I understood in each different case. 

So for the moment I will add case up to 30. Do you think I should add a version without template to be sure there is no error coming from the C++ code in case above 30?

Reply:
I suspect there are rapidly diminishing returns above N=10.  So I suggest having template cases up that that, and then a non-template for anything larger.

Reply:
Also, to make your life easier to have both kinds, you can use `Eigen::Dynamic` for when N>10, and also pass `n` as a variable that you use for the loops.  Then `N` can be only used for the Eigen template calls.

Reply:
Ok, so I increased the template to handle up to case 30 and I followed your suggestions when the case is above 30.